File: packages/auth-react-native/src/storage.ts
/**
 * Where a native app keeps its session.
 *
 * The session cookie is a bearer credential: anything holding it IS the signed-in
 * user until it expires. On a phone that means the OS keychain / keystore, not
 * AsyncStorage, which is plain unencrypted files any process with the sandbox
 * can read. `@authowl/expo` supplies an `expo-secure-store` adapter; bare React
 * Native apps typically use `react-native-keychain`.
 */
export interface SecureStorage {
  getItem(key: string): Promise<string | null>;
  setItem(key: string, value: string): Promise<void>;
  removeItem(key: string): Promise<void>;
}

/**
 * An in-memory store, for tests and for previews where persistence is unwanted.
 *
 * Explicitly NOT a default: silently falling back to memory would make sign-in
 * appear to work and then drop the session on the next app launch, which is a
 * confusing bug to chase. Callers must choose their storage.
 */
export class MemoryStorage implements SecureStorage {
  private readonly entries = new Map<string, string>();

  async getItem(key: string): Promise<string | null> {
    return this.entries.get(key) ?? null;
  }

  async setItem(key: string, value: string): Promise<void> {
    this.entries.set(key, value);
  }

  async removeItem(key: string): Promise<void> {
    this.entries.delete(key);
  }
}

File: sdks/flutter/lib/src/client/storage.dart
/// Where an AuthOwl client keeps its session.
library;

/// Persistent storage for the session cookie.
///
/// The session cookie is a bearer credential, so on a device this must be the
/// OS keychain (`flutter_secure_storage`), never `SharedPreferences`, which is
/// unencrypted files readable by anything inside the app sandbox.
///
/// Deliberately an interface with no default: silently falling back to memory
/// would make sign-in appear to work and then drop the session on the next app
/// launch, which is a confusing bug to chase.
abstract class AuthOwlStorage {
  Future<String?> read(String key);
  Future<void> write(String key, String value);
  Future<void> delete(String key);
}

/// An in-memory store, for tests and previews where persistence is unwanted.
class InMemoryAuthOwlStorage implements AuthOwlStorage {
  final Map<String, String> _entries = {};

  @override
  Future<String?> read(String key) async => _entries[key];

  @override
  Future<void> write(String key, String value) async => _entries[key] = value;

  @override
  Future<void> delete(String key) async => _entries.remove(key);
}
